### PR TITLE
common/hexutil: implement TextMarshaler, TextUnmarshaler

### PR DESCRIPTION
**Pull request by fjl**
This PR makes the wrapper types more generally applicable.
encoding.TextMarshaler is supported by most codec implementations (e.g.
for yaml).

The tests now ensure that package json actually recognizes the custom
marshaler implementation irrespective of how it is implemented.

The Uint type has new tests, too. These are tricky because uint size
depends on the CPU word size. Turns out that there was one incorrect
case where decoding returned ErrUint64Range instead of ErrUintRange

### CODE REVIEW

File: common/hexutil/json.go [hunk 1]
```diff
@@ -64,17 +70,11 @@ func (b Bytes) String() string {
...
-//
-//     type Foo [8]byte
-//
-//     func (f *Foo) UnmarshalJSON(input []byte) error {
-//         return hexutil.UnmarshalJSON("Foo", input, f[:])
-//     }
-func UnmarshalJSON(typname string, input, out []byte) error {
-	raw, err := checkJSON(input)
+// UnmarshalText method for fixed-size types.
+func UnmarshalFixedText(typname string, input, out []byte) error {
```
Comment (karalabe) [inline common/hexutil/json.go]: Couldn't we get the type name using reflection?

File: common/hexutil/json_test.go [hunk 2]
```diff
@@ -57,7 +59,7 @@ func referenceBytes(s string) []byte {
 
 var unmarshalBytesTests = []unmarshalTest{
 	// invalid encoding
-	{input: "", wantErr: errNonString},
+	{input: "", wantErr: errors.New("unexpected end of JSON input")},
```
Comment (karalabe) [inline common/hexutil/json_test.go]: Would be nice to put this error in its own variable. You're reusing it in quite a few places.

File: common/hexutil/hexutil_test.go [hunk 3]
```diff
@@ -28,9 +28,10 @@ type marshalTest struct {
 }
 
 type unmarshalTest struct {
-	input   string
-	want    interface{}
-	wantErr error
+	input        string
+	want         interface{}
+	wantErr      error // if set, decoding must fail on any platform
+	wantErr32bit error // if set, decoding must fail on 32bit platforms (used for Uint tests)
```
Comment (karalabe) [inline common/hexutil/hexutil_test.go]: I'm not entirely sure about this. Given that there may be errors working with int/uint, why support it at all? Seems to me to be just a fancy use case that will bite people.

File: common/hexutil/json.go (hunk 1 above)
Comment (fjl) [inline common/hexutil/json.go]: Yes, but then common/hexutil would need reflect. I'm also considering to pass the name in lower case at call sites.

File: common/hexutil/hexutil_test.go (hunk 3 above)
Comment (fjl) [inline common/hexutil/hexutil_test.go]: `uint` size depends on target arch, it's a fact of life. 

Let's say you have a (stdlib) API that expects a `uint` parameter. `hexutil.Uint` is nice if you want to decode a number from user input and use it with the API. The decoder does the range check for you.

The alternative would be to use `hexutil.Uint64` for decoding and then convert to `uint` when calling the API. To be safe against overflow, you'd need a check against `math.MaxUint`. Nobody would write those checks, so there'd just be a silent overflow.

Comment (karalabe) [inline common/hexutil/hexutil_test.go]: Fair enough.

File: common/hexutil/json.go (hunk 1 above)
Comment (karalabe) [inline common/hexutil/json.go]: May not be a huge overhead if it's only on the error path though.

File: common/hexutil/json_test.go (hunk 2 above)
Comment (karalabe) [inline common/hexutil/json_test.go]: @fjl Are you addressing this one or should I just merge?

Comment (fjl) [inline common/hexutil/json_test.go]: done

**Pull request merged**